### PR TITLE
Add ASHRAE Standard 205 Chiller object 

### PR DESCRIPTION
**Pull request by tanaya-mankad**
Pull request overview
---------------------
This branch will add the ability to read and process ASHRAE 205 representations for chillers (RS0001). See [NFP ASHRAE 205 Chillers](https://github.com/bigladder/EnergyPlus/blob/nfp-ashrae205-chillers/design/FY2022/NFP-ASHRAE205-Chillers.md)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nealkruis** [comment]
@EnergyArchmage, following up on our discussion during the technicalities call on this subject: it seems to me that the IDD inputs for "reference" flow rates on the EIR chillers are really "maximum requested" flow rates. For "constant flow" or "not modulated" chiller flow mode types, these are used directly for the requested flow and the temperature delta will be determined by the current capacity. For "leaving setpoint modulated" chillers, the requested flow is based on the current capacity and the setpoint temperature delta, but cannot exceed this value.

I'm trying to understand the implications of the [Chiller Flow Mode Type](https://bigladdersoftware.com/epx/docs/22-1/input-output-reference/group-plant-equipment.html#field-chiller-flow-mode-type) input. I think it would be helpful to have some real world examples of these configurations. For example, what does it mean if you have a "constant flow" chiller flow mode on a loop with a constant flow pump and the requested flow is less than the flow from the pump?

I really don't know a whole lot about chilled water loop design, but from what I gather from this [article on sequences-of-operations](https://www.esmagazine.com/articles/102254-the-three-stages-to-controlling-a-chiller-and-its-primary-secondary-pumps), it seems like the flow through the chillers is determined by the pump, and not the chiller itself.

I think we can add fields for the "maximum requested flow rates" on both the evaporator and condenser sides, but I feel like there might be a disconnect here between what is being described in the model and what is designed in reality (or maybe I'm missing something still?).

**Comment by EnergyArchmage** [comment]
Yes, those flow rates are maximum flows, I think of them as "Design Flow Rate".  I don't like the use of "reference" in the field names. I prefer names like "Design Heat Recovery Water Flow Rate." 

Generally the pump autosizing is done by summing up the component design flow rates.  If for some reason the pump is too small, then the plant components can be starved of flow and only get what the plant can deliver.  

EnergyPlus plant model is a simplification of what real plants would do.  Just trying to account for the major energy use of roughly equivalent system.  The control is different in the real world, because they have short timeframe controllers; they don't have to deal with one hour timesteps.  So in E+ the plant components request flows and the pumps follow the requests as best they can.

Usually it is the chilled water cooling coils on the loop's demand side that are the primary drivers of flow, when to turn on the plant etc.

**Comment by nealkruis** [comment]
To me "Design" implies something used for sizing. I imagine these flow rates are used in sizing routines, but they are also used in the main calculation method. It is certainly better than "reference". Do we want to do that here, or is it better to be consistent with the other objects even if it is not the best name?

**Comment by mjwitte** [comment]
> I'm trying to understand the implications of the [Chiller Flow Mode Type](https://bigladdersoftware.com/epx/docs/22-1/input-output-reference/group-plant-equipment.html#field-chiller-flow-mode-type) input. I think it would be helpful to have some real world examples of these configurations. For example, what does it mean if you have a "constant flow" chiller flow mode on a loop with a constant flow pump and the requested flow is less than the flow from the pump?

To answer this question. If there is a supply-side bypass in the model, then the constant flow chiller will get exactly the flow it asks for (which is the Reference flow rate for Chiller:Electric:EIR) and the excess flow will go through the bypass.  As @EnergyArchmage said earlier, if the pump delivers less than the request, the bypass will have zero flow and the chiller will get less that it requested.

Where the real plant would have control valves and logic for controlling the pumps, we don't have flow controls like that in EnergyPlus plant loops, so it's been wrapped into the chiller/boiler models instead for convenience. Perhaps it would be clearer if the field was named "Evaporator Flow Control Type"?  And maybe there should be a separate field for "Condenser Flow Control Type"?

File: CMakeLists.txt
Comment (Myoldmopar) [inline CMakeLists.txt:347]: :+1: No need to run these each time.

File: design/FY2022/NFP-ASHRAE205-Chillers.plantuml
Comment (Myoldmopar) [inline design/FY2022/NFP-ASHRAE205-Chillers.plantuml:1]: I guess GitHub won't render plantuml here in the PR, too bad.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in]: You have min fields set to 12 here.  The 12th field is _Chilled Water Maximum Requested Flow Rate_, which does not have a `\required-field` attribute.  So is the min fields 11?  Or is the 12th field required?

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc:131]: :+1:

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc]: This is being created as an `int`, so the `auto` shorthand is really not doing anything valuable here.  Probably not worth changing if there's no other changes.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc:162]: `if (!thisChiller.Representation)`?  Not a big deal, but it feels a little like `if (false == someFlag)`.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc]: This block shouldn't be in here.  The `setOutputVariables` function should be called from the `oneTimeInit` function, and you should eliminate the `->oneTimeFlag` variable and check entirely.  The point of the `oneTimeInit` was to eliminate this type check, and replace it with an intentional one-time pass over all the plant components [here](https://github.com/NREL/EnergyPlus/blob/a1211bbb5613f31be3bc6b789d48652b956b8005/src/EnergyPlus/Plant/PlantManager.cc#L1818-L1835), right after the factory methods are called.  I think this change should be made before merging.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc]: Should you issue a warning?

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc:1451]: SolveRoot actually takes a `Function` and generic payload [now](https://github.com/NREL/EnergyPlus/blob/a1211bbb5613f31be3bc6b789d48652b956b8005/src/EnergyPlus/General.hh#L85), but we've not really exercised it that way.  Would have been a nice opportunity to use a lambda and generic payload on some new modern C++ code in here, but oh well.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc:1643]: Interesting.  MinLoad is usually 0.

File: src/EnergyPlus/ChillerElectricASHRAE205.hh
Comment (Myoldmopar) [inline src/EnergyPlus/ChillerElectricASHRAE205.hh:169]: :+1:

File: src/EnergyPlus/Data/EnergyPlusData.hh
Comment (Myoldmopar) [inline src/EnergyPlus/Data/EnergyPlusData.hh:1]: Why is this file showing all the lines changed?  Did something break with line endings?

File: src/EnergyPlus/Plant/Component.hh
Comment (Myoldmopar) [inline src/EnergyPlus/Plant/Component.hh:162]: Yeah, this is indeed silly out here.

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
This looks really close actually.  I had a few comments, but I think only a couple actually need to be addressed for merge.

File: src/EnergyPlus/Data/EnergyPlusData.hh
Comment (Myoldmopar) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: (please address this one)

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc]: (please address this one)

File: idd/Energy+.idd.in
Comment (Myoldmopar) [reply idd/Energy+.idd.in]: (please address this one)

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (tanaya-mankad) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:162]: If `if (!stuff)` is preferred, I'll change this - but since C++11 introduced `nullptr_t` as its own type, I tend to use it explicitly rather than relying on implicit "falsy" conversions.
Also happy to move the order of comparison around if this is not the E+ way. (I waffle on which I like better.)

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (tanaya-mankad) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc]: Oops, meant to remove that check. Fixed.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (tanaya-mankad) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc]: Fixed. 
So the way we had it, oneTimeInit was probably being called two times? 
A bit of boilerplate component code in the wiki might be helpful for this - if an unsuspecting dev copies and pastes from an existing file, the error gets propagated up to the reviewer every time. :(

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (tanaya-mankad) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1451]: Bummer. Next time, or a subsequent revision then?
I also wonder if the template could use SFINAE rather than static_assert? (idle thought)

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (tanaya-mankad) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc]: Fixed.

File: idd/Energy+.idd.in
Comment (tanaya-mankad) [reply idd/Energy+.idd.in]: Fixed (min=11).

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (nealkruis) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1643]: Maybe it should be? I was following the lead of the other chiller objects which use rated capacity times minimum part load ratio. Though it looks like a typo here. `this->MaxSequenceNumber` should be `this->MinSequenceNumber`. That's my bad. @tanaya-bigladder, can you make that change if you already have other modifications in progress?

File: src/EnergyPlus/Data/EnergyPlusData.hh
Comment (nealkruis) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: I noticed this too. It actually happened after pulling in develop. I had to manually merge conflicts that really shouldn't have been required.

Comment (tanaya-mankad) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: I tried to fix this yesterday but it appears it hasn't stuck. In one file comparison tool, I noticed that line endings were `0x 0D 0A` on our manually merged copy, and `0x 0A `on develop, but on another tool I see `CRLF` on both. Yesterday's check in should have made them consistently 0x 0A, but it's still showing as a diff...

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (EnergyArchmage) [inline src/EnergyPlus/ChillerElectricASHRAE205.cc]: WaterHeaterMixed ? looks like copy/paste error

File: src/EnergyPlus/Data/EnergyPlusData.hh
Comment (Myoldmopar) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: Do you want me to take a shot at fixing the issue?  I'm happy to try if I have write access to this branch.

Comment (tanaya-mankad) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: From looking at a few different files, I deduce that the "correct" line endings are expected to be just `LF`. The version of [EnergyPlusData.hh on the develop branch](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/Data/EnergyPlusData.hh) uses `CRLF`. It appears to be a lone wolf.

Comment (nealkruis) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: If it helps, this popped up when I merged in the change that removed the PackagedTerminalAirConditioner (or HeatPump?) from the list.

Comment (Myoldmopar) [reply src/EnergyPlus/Data/EnergyPlusData.hh:1]: Yeah, it's showing up as CRLF on my develop checkout as well.  But in this branch the line endings are normalized back.  This is a good correction so just thanks and carry on!

**Comment by Myoldmopar** [comment]
Thanks for making the changes and helping resolve the file line endings confusion.  @EnergyArchmage made a comment about a possible typo that should be fixed or resolved in some way.  There are some integration failures because of separate PRs going in.  The Coil:*:WaterToAirHeatPump:EquationFit objects in your new IDFs in this branch need transition.  Also, I think your input files are showing version 22.1 still, so go ahead and update that as well.  Overall I think this is getting very close though.

**Comment by mjwitte** [comment]
@tanaya-bigladder Is this ready for final review?

**Comment by tanaya-mankad** [comment]
@mjwitte, it is!
As @nealkruis mentioned in the iteration meeting Wednesday, we'll likely update a handful of internal features that do not affect IO freeze, but it will be done in a separate PR.

**Comment by lymereJ** [comment]
@tanaya-bigladder @nealkruis - Sorry for only bringing this up now (and perhaps you've addressed it and missed it when skimming through the proposed changes), are you not planning on having the ASHRAE205 chillers report their IPLV?

**Comment by nealkruis** [comment]
@lymereJ we originally did not include this in scope partly because the ASHRAE 205 object is substantially different in structure from the curve-based chillers that use `CalcChillerIPLV`. After looking at that function a bit more, I feel like a good follow up task would be to generalize it as a member method of a chiller base class and eliminate some of the redundant code used to calculate the IPLV.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:162]: I would say cases like this, the "not null" check, are one type of truthy checks that I actually prefer over being explicit and comparing it to null.  But it's nothing to hold up the PR.  Moving on.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1451]: Next time.

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1643]: I will rely on you all here as you know the 205 model better than me.  I really would assume it would be 0 for minimum load, but if you think minimum load should be treated this way then so be it.

File: src/EnergyPlus/ChillerElectricASHRAE205.hh
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerElectricASHRAE205.hh:169]: I am surprised you didn't use an `EPVector` or a `std::vector` to hold the chillers.  That's another missed opportunity.

**Review by Myoldmopar — APPROVED** [review]
I'm happy with these changes now.  There are a few things that would be nice to cleanup post IO freeze, but let's get this one going in.  I'm testing locally and I need to check out CI as well next.

File: src/EnergyPlus/ChillerElectricASHRAE205.hh
Comment (nealkruis) [reply src/EnergyPlus/ChillerElectricASHRAE205.hh:169]: This is pretty easy to change I imagine. I think the Array1D was just mimicked from the other chiller source files.

Comment (Myoldmopar) [reply src/EnergyPlus/ChillerElectricASHRAE205.hh:169]: It is a pretty easy change to make, but don't make it here.  I'm doing final testing locally, no need to waste CI cycles on this change. 

And yeah, it still shows up in some components, but as we refactor, we are trying hard to move to _at least_ `EPVector`, and ideally just `std::vector` for runtime-sized lists.  The change to `EPVector` would almost certainly require zero other changes.

**Comment by Myoldmopar** [comment]
Everything passes fine locally.  Looking at the new detailed IDF error file it has this, are you expecting this type of output??

```
   ** Warning ** GetDaylightingControls: Fraction of zone or space controlled by the Daylighting reference points is < 1.0.
   **   ~~~   ** ..discovered in Daylighting:Controls="PERIMETER_TOP_ZN_4_DAYLCTRL", only 0.524 of the zone or space is controlled.
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid axis with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid axis with size 4
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid axis with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid axis with size 4
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding (int) grid axis with size 4
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid axis with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller1": Adding grid table with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid axis with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid axis with size 4
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid axis with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid axis with size 4
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding (int) grid axis with size 4
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 64
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid axis with size 1
   ************* Chiller:Electric:ASHRAE205 "CoolSys1 Chiller2": Adding grid table with size 1
   ** Warning ** Calculated design heating load for zone=BASEMENT is zero.
   **   ~~~   ** Check Sizing:Zone and ZoneControl:Thermostat inputs.
   ** Warning ** Calculated design heating load for zone=CORE_BOTTOM is zero.
```

File: src/EnergyPlus/ChillerElectricASHRAE205.cc
Comment (nealkruis) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1643]: I honestly don't know how I should be interpreting `MinLoad`. ASHRAE 205 chillers can cycle down to zero just like other chillers can below their minimum part load ratio, but the [EIR chiller doesn't use zero](https://github.com/NREL/EnergyPlus/blob/2b7eae49e2f7e8401a3904852cc200fbe87185f7/src/EnergyPlus/ChillerElectricEIR.cc#L185). Maybe they both should?

Comment (mjwitte) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1643]: My understanding is that `MinLoad` is only used for load allocation when there is more than one piece of equipment available. The EIR chiller `MinPartLoadRat` is a user input that defaults to 0.1. The IDD describes it as: Part load ratio below which the chiller starts cycling on/off to meet the load. So the goal is for load control to avoid putting one chiller in cycling when another one is also running. So if you have a way to search the performance map to find where it switches from continuous to cycling, then use that. Otherwise I would go with zero.

Comment (nealkruis) [reply src/EnergyPlus/ChillerElectricASHRAE205.cc:1643]: Thanks, @mjwitte! The ASHRAE 205 performance map is fairly analogous to the multi-/variable-speed DX coils. It defines performance at several compressor speeds and cycles below the lowest speed. Based on what you have described, I believe we are handling this correctly.

**Comment by nealkruis** [comment]
@Myoldmopar those warnings are testing remnants that should have been cleaned out. Would you like us to do that now, or on the follow up PR?

**Comment by Myoldmopar** [comment]
Just do it in the follow-up branch, this is fine to go in.  Other than those warnings things look good here.  Thanks @nealkruis and @tanaya-bigladder .

**Pull request merged**
